File: c++/cote/LTCD/세_수의_합.cpp
#include <iostream>
#include <random>
#include <vector>

#define TEST_CASE 10
#define N 10000

using namespace std;


vector<int*> solve(int* nums, int n){
	
	vector<int*> ret;
	sort(nums, nums+n);

	for(int i=0; i<n; i++){

		int lo = i+1; 
		int hi = n-1;

		while(lo < hi){
			int sum = nums[i] + nums[lo] + nums[hi];
			if(sum > 0){
				hi--;
			}
			if(sum < 0){
				lo++;
				continue;
			}
			if(sum == 0){
				ret.push_back(new int[3]{nums[i], nums[lo], nums[hi]});
				while(lo < hi &&nums[hi-1] == nums[hi]){
					hi--;
				}
				while(lo < hi && nums[lo +1] == nums[lo]){
					lo++;
				}

				lo++;
				hi--;
			}
		}

	}

	return ret;
}

int main(void){

	for(int test_case=0; test_case<TEST_CASE; test_case++){
		default_random_engine gen(43);
		uniform_int_distribution<int> dis(-N,N);

		int n = gen() % (N-100) + 100;
		int* nums = new int[n];

		for(int i=0; i<n; i++){
			nums[i] = dis(gen);
		}

		vector<int*> ret = solve(nums, n);
		cout << "[";
		for(int i=0; i<ret.size(); i++){
			cout <<"[";
			for(int j=0; j<3; j++){
				cout << ret[i][j]<< ",";
			}	
			cout <<"], ";
		}
		cout <<"]" << endl;

		delete[] nums;
	}

	return 0;
}

File: python/cote/LTCD/세_수의_합.py
from typing import *
import random

TEST_CASES = 10
N = 10000


def search(target: int, nums: List[int]) -> List[List[int]]:
    ret = []

    left, right = target + 1, len(nums) - 1

    while left < right:
        sum = nums[target] + nums[left] + nums[right]

        if sum < 0:
            left += 1
        elif sum > 0:
            right -= 1
        else:
            ret.append(sorted([nums[left], nums[right], nums[target]]))

            while left < right and nums[left] == nums[left + 1]:
                left += 1
            while left < right and nums[right] == nums[right - 1]:
                right -= 1

            left += 1
            right -= 1

    return ret


def solve(nums: List[int]) -> List[List[int]]:
    ret = []

    nums.sort()

    for i in range(len(nums) - 2):

        if nums[i] > 0 and nums[i] == nums[i - 1]:
            continue

        tmp = search(i, nums)

        if len(tmp) != 0:
            ret.extend(tmp)

    return ret


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):
        n = random.randrange(100, N)
        nums = random.sample(range(-(n // 2), n // 2), n // 2)
        print(solve(nums))
